rue89-tech/reg-ext-form
# [BB-1493] Implement a custom registration form

This implements a custom registration form for the Rue89 Open edX instance

**Testing instructions**:
* Clone this repo in a location accessible under the Ironwood devstack.
* Navigate inside the top-level directory of the repository inside the devstack and run `pip install -e .` to install the app in editable mode.
* Add `reg_ext_form` to the `INSTALLED_APPS`.
* Run `python manage.py lms migrate reg_ext_form` to apply the migrations and create the model.
* ~~Add the following JSON to the `SiteConfiguaration` model for the default site.~~ Since there is a bug in the upstream source code, add these in `lms.env.json` instead.
  ```javascript
  {
    "ENABLE_COMBINED_LOGIN_REGISTRATION":true,
    "REGISTRATION_EXTRA_FIELDS":{
      "city":"required",
      "gender":"required",
      "year_of_birth":"required",
      "level_of_education":"required",
      "goals":"hidden",
      "honor_code":"hidden",
      "country":"required",
      "mailing_address":"hidden",
      "terms_of_service":"hidden"
    }
  }
  ```
* Add a key `REGISTRATION_EXTENSION_FORM` to `lms.env.json` with the dotted path to the `ExtraInfoForm` as its value - `reg_ext_form.forms.ExtraInfoForm`.
* Restart LMS.
* Open the LMS and navigate to the registration page.
* Verify that the checkbox with the correct text is present and is required for registration.
* Also verify that the model can be viewed via the Django admin interface.

**Reviewer**: @xitij2000

**Author notes and concerns**:
* Since the translated text is not provided for the error messages, the default French error message for "This field is required" is used. I tried to create a custom error message based on the field but since I ran into issues with `'` in the message getting encoded as a HTML entity even after marking it safe using `mark_safe`, I decided not to pursue it further considering the time constraints for this ticket.
* The position of the field is in the middle of the form and the only way to push it down to the bottom is to tweak multiple JS files in the theme to override the built-in files. We don't have the time budget to do it as a part of this ticket.
